src/dragonball/src/resource_manager.rs

```diff
@@ -0,0 +1,683 @@
+        // if shared_irq is used, just return the shared irq num.
+        if shared {
+            return Some(SHARED_IRQ);
+        }
+
+        let mut constraint = Constraint::new(1u32);
+        if let Some(v) = fixed {
+            if v == SHARED_IRQ {
+                return None;
+            }
+
+            constraint.min = v as u64;
+            constraint.max = v as u64;
+        }
+        // Safe to unwrap() because we don't expect poisoned lock here.
+        let mut legacy_irq_pool = self.legacy_irq_pool.lock().unwrap();
+        let key = legacy_irq_pool.allocate(&constraint);
+        if let Some(k) = key.as_ref() {
+            legacy_irq_pool.update(k, ());
+        }
+        key.map(|v| v.min as u32)
+    }
+
+    /// Free a legacy irq number.
+    ///
+    /// Panic if the irq number is invalid.
+    pub fn free_legacy_irq(&self, irq: u32) {
+        // if the irq number is shared_irq, we don't need to do anything.
+        if irq == SHARED_IRQ {
+            return;
+        }
+
+        if !(LEGACY_IRQ_BASE..=LEGACY_IRQ_MAX).contains(&irq) {
+            panic!("invalid irq number when freeing legacy irq");
+        }
+        let key = Range::new(irq, irq);
+        // Safe to unwrap() because we don't expect poisoned lock here.
+        self.legacy_irq_pool.lock().unwrap().free(&key);
+    }
+
+    /// Allocate a group of MSI irq numbers.
+    ///
+    /// The allocated MSI irq numbers may or may not be naturally aligned.
+    pub fn allocate_msi_irq(&self, count: u32) -> Option<u32> {
+        let constraint = Constraint::new(count);
+        // Safe to unwrap() because we don't expect poisoned lock here.
+        let mut msi_irq_pool = self.msi_irq_pool.lock().unwrap();
+        let key = msi_irq_pool.allocate(&constraint);
+        if let Some(k) = key.as_ref() {
+            msi_irq_pool.update(k, ());
+        }
+        key.map(|v| v.min as u32)
+    }
+
+    /// Allocate a group of MSI irq numbers, naturally aligned to `count`.
+    ///
+    /// This may be used to support PCI MSI, which requires the allocated irq number is naturally
+    /// aligned.
+    pub fn allocate_msi_irq_aligned(&self, count: u32) -> Option<u32> {
+        let constraint = Constraint::new(count).align(count);
+        // Safe to unwrap() because we don't expect poisoned lock here.
+        let mut msi_irq_pool = self.msi_irq_pool.lock().unwrap();
+        let key = msi_irq_pool.allocate(&constraint);
+        if let Some(k) = key.as_ref() {
+            msi_irq_pool.update(k, ());
+        }
+        key.map(|v| v.min as u32)
+    }
+
+    /// Free a group of MSI irq numbers.
+    ///
+    /// Panic if `irq` or `count` is invalid.
+    pub fn free_msi_irq(&self, irq: u32, count: u32) {
+        if irq < MSI_IRQ_BASE
+            || count == 0
+            || irq.checked_add(count).is_none()
+            || irq + count - 1 > MSI_IRQ_MAX
+        {
+            panic!("invalid irq number when freeing legacy irq");
```

Comment:
Again, please add a unit test using `#[should_panic] ` for this and any other functions that panic.

Reply:
The panic has been removed.